### PR TITLE
Introduce AsyncWriter which batches and transmits entries in the background

### PR DESCRIPTION
**Pull request by dazuma**
This is an attempt to address issue #898 by providing a way for Logger to make non-blocking calls. Please feel free to critique the approach. The idea is to provide an object that queues up log entries and writes them in batches from a background thread.

### CODE REVIEW

**Comment by dazuma** [comment]
@blowmage @quartzmo Wanted to see what feedback you have on this.

So far I've heard from Mike that: currently the Logger still uses blocking calls by default, and you have to opt-in to use AsyncWriter's non-blocking calls. Sounds like Mike would rather Logger use non-blocking by default. That SGTM, I can do that. (Project#logger would by default first create an AsyncWriter and then use it to construct the Logger.)

Anything else?

**Comment by blowmage** [comment]
I do think this should be the default behavior of the Ruby Logger provided by Google Cloud Logging. I can't think of a reason why we would not want this, as long as it doesn't load GRPC into a rails app's main thread. Seems like a win to me.

**Comment by quartzmo** [comment]
@dazuma Sorry for the delay in responding. I also like the idea of the Logger being non-blocking by default, since it sounds like the right choice for most users. For cases in which it is not the right choice, what do you think about adding documentation that discusses environments in which the background thread likely **won't** work, and what steps (if any) must be taken to switch to blocking calls (or some other possible workaround)?

For example, I'm wondering about deployments on Unicorn (see [Implementing worker threads in Rails - Forking Servers](http://blog.arkency.com/2013/06/implementing-worker-threads-in-rails/#forking_servers).) But is this even a valid scenario? What are the deployment scenarios in which application might want to write to Stackdriver Logging?

**Comment by dazuma** [comment]
Thanks Mike and Chris. I'll update the PR to make Logger use non-blocking calls by default (unless explicitly configured otherwise). I'll also investigate the forking issue and come up with a resolution. Yes, using Unicorn is a valid use case, as are threaded servers like Puma.

**Comment by quartzmo** [comment]
There are some interesting errors in the Travis CI build, linked above.

For example:

```sh
  1) Error:
Google::Cloud::Logging::AsyncWriter::mock_logging#test_0001_writes a single entry:
ThreadError: current thread not owner
    /Users/travis/.rvm/rubies/ruby-2.1.9/lib/ruby/2.1.0/monitor.rb:246:in `mon_check_owner'
    /Users/travis/.rvm/rubies/ruby-2.1.9/lib/ruby/2.1.0/monitor.rb:147:in `broadcast'
    /Users/travis/build/GoogleCloudPlatform/google-cloud-ruby/google-cloud-logging/lib/google/cloud/logging/async_writer.rb:150:in `block in write_entries'
    /Users/travis/.rvm/rubies/ruby-2.1.9/lib/ruby/2.1.0/monitor.rb:211:in `mon_synchronize'
    /Users/travis/build/GoogleCloudPlatform/google-cloud-ruby/google-cloud-logging/lib/google/cloud/logging/async_writer.rb:143:in `write_entries'
    /Users/travis/build/GoogleCloudPlatform/google-cloud-ruby/google-cloud-logging/test/google/cloud/logging/async_writer_test.rb:68:in `block (2 levels) in <top (required)>'
```

**Comment by dazuma** [comment]
@quartzmo Yeah I noticed. Turns out I'd used the wrong method to test thread status. Should be fixed now.

**Comment by quartzmo** [comment]
Just wondering (don't consider this a request by any means), but would it be possible and would there be any value in a mechanism that would force the AsyncWriter to block while it writes its entire queue to the backend? Something akin to `flush`...

**Comment by quartzmo** [comment]
Do you want to add a sentence or two about the background thread to [the Logging guide section on writing entries](https://github.com/GoogleCloudPlatform/google-cloud-ruby/blob/master/google-cloud-logging/lib/google/cloud/logging.rb#L198-L248)?

**Comment by dazuma** [comment]
Re: an `AsyncWriter#flush` call. It's possible, but there are several possible semantics, and without some use cases it's not yet clear to me which is the useful one.

For example: If the semantics are that the queue is empty when the call completes, then it would have to block not only the call itself, but also any other concurrent calls to write_entries (which could otherwise cause the call to take an arbitrary amount of time by continually adding to the queue.) Alternatively, we could get around that issue by weakening the guarantee: for example, that only the entries added previously by the same thread have been flushed.

I'll take a look at the guide.

**Comment by quartzmo** [comment]
Re: `AsyncWriter#flush`, sounds good, I just wanted to bring up the idea since it occurred to me. I don't think it needs to be done for this PR. Thanks for the thoughtful answer!

**Comment by quartzmo** [comment]
@dazuma Hmmm, your additions to the guide make perfect sense, but I now wonder if the existing content needs to be revised. The problem is that under the heading "Writing log entries", we offer `logging.write_entries entry` as the preferred approach. However, this now feels like a lower-level usage that gets underneath the new preferred mechanism of the `AsyncWriter`. The following section, "Creating a Ruby Logger implementation", is more in the spirit of easy async logging that just works. Should we move this section above "Writing log entries", and begin it with the vanilla (canonical) async example, then discuss setting `async_writer: false`, and then follow with the lower-level discussion about `write_entries`?

If this is a bit hard to follow, I can easily provide a gist with the edits.

**Comment by dazuma** [comment]
@quartzmo I'm actually inclined to leave the guide's flow as it is, with AsyncWriter and the Logger implementation as secondary to the "raw" API. My reasoning:

* All the other google-cloud-* libraries are focused on being a client for the API first. Logging is a bit special in that it _also_ offers integration with the standard Ruby logging "framework" (i.e. the Logger class). However, that capability could arguably be more closely connected with what we're doing on the stackdriver side (i.e. integrating with Rails and other frameworks). It doesn't fit as well with the overall paradigm of being an API client, so I'm hesitant to put it front and center in this library.
* AsyncWriter, although it's perhaps "preferred" in that it's non-blocking, still involves increased complexity over `logging.write_entries` simply because you have to think a bit about concurrency (e.g. the fact that you don't get errors back from your API calls.) It also covers only a subset of the API (write_entries only, not any of the query methods). As such, I also think it's better considered an optional add-on rather than putting it first.

That said, maybe just one or two sentences at the top of the guide to the effect of "if you just want to log to the google logging service, consider the Logger class described further down; otherwise read on to learn about the API client" would accomplish what you're looking for? What do you think?

**Comment by quartzmo** [comment]
@dazuma Makes sense, I think you're right! We can keep the focus on the simpler, lower-level usage, and then explain the more complex options. Adding a couple of opening sentences to provide context sounds good.

**Comment by dazuma** [comment]
Sounds good, I'll make those additions.

**Comment by quartzmo** [comment]
LGTM! I'll give @blowmage the chance to review and merge, however.

**Comment by quartzmo** [comment]
Thanks again @dazuma, this is really impressive work!!!
